`aws_write_only_arguments`: recommend write-only arguments where available (#860)
terraform-linters/tflint-ruleset-aws: docs/rules/aws_write_only_attributes.md

# aws_write_only_attributes

> Disallows using the normal attribute containing sensitive information, in favor of an available write-only attribute. This is only valid for Terraform version 1.11.0 and later. If there alternative options available to prevent this information ending up in state, these are suggested as well.

Reviewer: ```suggestion
Recommends using available [write-only arguments](https://developer.hashicorp.com/terraform/language/resources/ephemeral/write-only) instead of the original sensitive attribute. This is only valid for Terraform v1.11+.
```

## Example

> In this example `aws_secretsmanager_secret_version` is used, but you can replace that with any resource with write-only attributes:

Reviewer: ```suggestion
This example uses `aws_secretsmanager_secret_version`, but the rule applies to all resources with write-only attributes:
```

```hcl
resource "aws_secretsmanager_secret_version" "test" {
  secret_string = var.secret
}
```

```
$ tflint
1 issue(s) found:

Warning: [Fixable] "secret_string" is a non-ephemeral attribute, which means this secret is stored in state. Please use "secret_string_wo". (aws_write_only_attributes)

  on test.tf line 3:
   3:   secret_string = var.secret

```

## Why

> Saving secrets to state or plan files is a bad practice. It can cause serious security issues. Keeping secrets from these files is possible in most of the cases by using write-only attributes.

Reviewer: ```suggestion
By default, sensitive attributes are still stored in state, just hidden from view in plan output. Other resources are able to refer to these attributes. Current versions of Terraform also include support for write-only arguments, which are not persisted to state. Other resources cannot refer to their values.

Using write-only arguments mitigates the risk of a malicious actor obtaining privileged credentials by accessing Terraform state files directly. Prefer using them over the original sensitive attribute unless you need to refer to it in other blocks, such as a [root `output`](https://developer.hashicorp.com/terraform/language/values/outputs#ephemeral-avoid-storing-values-in-state-or-plan-files), that cannot be ephemeral.
```

Security recommendations should be precise, on:

* The risk/threat
* The mitigation
* Any tradeoffs/limitations

Let the reader interpret how it applies to their code/system. Avoid subjective characterization, e.g.,:

* Bad practice
* "Serious" issues
* "possible in most of the cases"

## How To Fix

> Replace the `secret_string` with `secret_string_wo`, preferable via using an ephemeral resource as source or using an ephemeral variable:

Reviewer: ```suggestion
Replace the attribute with its write-only argument equivalent. Reference an ephemeral resource or ephemeral variable to ensure that the sensitive value is not persisted to state.
```
